### PR TITLE
Split Memo node state into Restoring and Computing

### PR DESCRIPTION
**Pull request by snowleopard**
This is a follow up to #4797, which not only brings the performance back but even improves it.

The main idea behind the optimisation is to split the `Considering` state into `Restoring` and `Computing`, which allows us to avoid creating `compute` computations, and associated stack frames, when they are unnecessary (e.g. in incremental zero rebuilds). Also, we no longer need to deduplicate call stacks when reporting cycles, because `restore_from_cache` and `compute` are no longer nested. Finally, by moving `run` into `Computation0` it became possible to save one indirection, since the `Restoring` constructor now stores a record with just a single field.

Below are some microbenchmarks. This PR improves the restoring from cache scenario by around 30% compared to #4797, and by around 15% compared to Memo before #4797.

```
┌────────────────────────────────────┬──────────────┬─────────────┬───────────┬───────────┬────────────┐
│ Do not reuse DAG nodes (this PR)   │     Time/Run │     mWd/Run │  mjWd/Run │  Prom/Run │ Percentage │
├────────────────────────────────────┼──────────────┼─────────────┼───────────┼───────────┼────────────┤
│ 1-bind (create and compute)        │     584.28ns │     821.00w │     0.14w │     0.14w │      0.34% │
│ 1-bind (incr and recompute)        │     597.01ns │     793.00w │     0.11w │     0.11w │      0.35% │
│ 1-bind (restore from cache)        │     378.37ns │     507.00w │           │           │      0.22% │
│ 20-reads (create and compute)      │   2_098.38ns │   3_182.00w │     0.78w │     0.78w │      1.22% │
│ 20-reads (incr and recompute)      │   1_962.54ns │   2_921.00w │     0.51w │     0.51w │      1.14% │
│ 20-reads (restore from cache)      │   1_745.11ns │   2_635.00w │     0.41w │     0.41w │      1.02% │
│ clique (create and compute)        │  64_065.26ns │  74_958.00w │   382.68w │   382.68w │     37.27% │
│ clique (incr and recompute)        │  62_583.16ns │  75_610.00w │   193.39w │   193.39w │     36.41% │
│ clique (restore from cache)        │  18_737.88ns │  27_690.00w │    41.55w │    41.55w │     10.90% │
│ bipartite (create and compute)     │ 171_891.32ns │ 171_871.00w │ 2_851.56w │ 2_851.56w │    100.00% │
│ bipartite (incr and recompute)     │ 124_778.29ns │ 138_315.00w │   365.31w │   365.31w │     72.59% │
│ bipartite (restore from cache)     │  34_508.46ns │  55_620.00w │    21.11w │    21.11w │     20.08% │
│ memo diamonds (create and compute) │  12_267.94ns │  15_871.00w │    44.92w │    44.92w │      7.14% │
│ memo diamonds (incr and recompute) │  12_789.34ns │  16_426.00w │    27.78w │    27.78w │      7.44% │
│ memo diamonds (restore from cache) │   4_871.63ns │   6_320.00w │     8.70w │     8.70w │      2.83% │
└────────────────────────────────────┴──────────────┴─────────────┴───────────┴───────────┴────────────┘

┌────────────────────────────────────┬──────────────┬─────────────┬───────────┬───────────┬────────────┐
│ Do not reuse DAG nodes (Dune main) │     Time/Run │     mWd/Run │  mjWd/Run │  Prom/Run │ Percentage │
├────────────────────────────────────┼──────────────┼─────────────┼───────────┼───────────┼────────────┤
│ 1-bind (create and compute)        │     571.79ns │     836.00w │     0.16w │     0.16w │      0.33% │
│ 1-bind (incr and recompute)        │     585.16ns │     808.00w │     0.14w │     0.14w │      0.34% │
│ 1-bind (restore from cache)        │     499.30ns │     689.00w │     0.11w │     0.11w │      0.29% │
│ 20-reads (create and compute)      │   2_094.54ns │   3_197.00w │     0.81w │     0.81w │      1.21% │
│ 20-reads (incr and recompute)      │   1_934.98ns │   2_936.00w │     0.53w │     0.53w │      1.12% │
│ 20-reads (restore from cache)      │   1_825.23ns │   2_817.00w │     0.49w │     0.49w │      1.05% │
│ clique (create and compute)        │  62_872.44ns │  75_314.00w │   399.13w │   399.13w │     36.29% │
│ clique (incr and recompute)        │  65_345.74ns │  75_966.00w │   336.78w │   336.78w │     37.72% │
│ clique (restore from cache)        │  23_539.43ns │  33_514.00w │    85.28w │    85.28w │     13.59% │
│ bipartite (create and compute)     │ 173_244.04ns │ 172_876.00w │ 2_886.32w │ 2_886.32w │    100.00% │
│ bipartite (incr and recompute)     │ 132_908.54ns │ 149_238.00w │   402.65w │   402.65w │     76.72% │
│ bipartite (restore from cache)     │  46_742.35ns │  72_182.00w │    40.85w │    40.85w │     26.98% │
│ memo diamonds (create and compute) │  12_344.34ns │  16_106.00w │    48.19w │    48.19w │      7.13% │
│ memo diamonds (incr and recompute) │  13_389.32ns │  16_661.00w │    44.56w │    44.56w │      7.73% │
│ memo diamonds (restore from cache) │   7_413.33ns │  10_142.00w │    22.46w │    22.46w │      4.28% │
└────────────────────────────────────┴──────────────┴─────────────┴───────────┴───────────┴────────────┘

┌────────────────────────────────────┬──────────────┬─────────────┬───────────┬───────────┬────────────┐
│ Reuse DAG node (before #4797)      │     Time/Run │     mWd/Run │  mjWd/Run │  Prom/Run │ Percentage │
├────────────────────────────────────┼──────────────┼─────────────┼───────────┼───────────┼────────────┤
│ 1-bind (create and compute)        │     564.43ns │     810.00w │     0.14w │     0.14w │      0.33% │
│ 1-bind (incr and recompute)        │     572.51ns │     782.00w │     0.12w │     0.12w │      0.34% │
│ 1-bind (restore from cache)        │     425.57ns │     581.00w │           │           │      0.25% │
│ 20-reads (create and compute)      │   2_127.96ns │   3_209.00w │     0.80w │     0.80w │      1.25% │
│ 20-reads (incr and recompute)      │   1_997.88ns │   2_948.00w │     0.52w │     0.52w │      1.17% │
│ 20-reads (restore from cache)      │   1_833.68ns │   2_747.00w │     0.43w │     0.43w │      1.07% │
│ clique (create and compute)        │  64_036.37ns │  75_536.00w │   391.45w │   391.45w │     37.53% │
│ clique (incr and recompute)        │  63_838.59ns │  76_250.00w │   238.90w │   238.90w │     37.42% │
│ clique (restore from cache)        │  21_152.35ns │  30_988.00w │    57.83w │    57.83w │     12.40% │
│ bipartite (create and compute)     │ 170_610.40ns │ 172_610.00w │ 2_866.05w │ 2_866.05w │    100.00% │
│ bipartite (incr and recompute)     │ 129_543.39ns │ 144_048.00w │   384.75w │   384.75w │     75.93% │
│ bipartite (restore from cache)     │  41_303.35ns │  64_094.00w │    28.37w │    28.37w │     24.21% │
│ memo diamonds (create and compute) │  12_347.91ns │  15_680.00w │    46.01w │    46.01w │      7.24% │
│ memo diamonds (incr and recompute) │  12_864.23ns │  16_275.00w │    33.97w │    33.97w │      7.54% │
│ memo diamonds (restore from cache) │   5_871.46ns │   7_914.00w │    13.56w │    13.56w │      3.44% │
└────────────────────────────────────┴──────────────┴─────────────┴───────────┴───────────┴────────────┘
```

### CODE REVIEW

**Comment by snowleopard** [comment]
The second commit makes another improvement by getting rid of stale computations and merging two parts of a node's state (`state` plus `last_cached_value`) into a single `state` with five possible values. I'm not 100% happy with the resulting code yet, so any ideas on how to make it better are welcome.

Microbenchmarks show 10-20% improvement for the scenarios that involve the compute phase. Building a large repository also shows that incremental zero rebuilds got faster by 10-15% compared to just the first commit.

```
┌────────────────────────────────────┬──────────────┬─────────────┬───────────┬───────────┬────────────┐
│ Five states                        │     Time/Run │     mWd/Run │  mjWd/Run │  Prom/Run │ Percentage │
├────────────────────────────────────┼──────────────┼─────────────┼───────────┼───────────┼────────────┤
│ 1-bind (create and compute)        │     445.62ns │     633.00w │           │           │      0.29% │
│ 1-bind (incr and recompute)        │     447.09ns │     606.00w │           │           │      0.29% │
│ 1-bind (restore from cache)        │     373.10ns │     514.00w │           │           │      0.24% │
│ 20-reads (create and compute)      │   1_986.52ns │   3_032.00w │     0.72w │     0.72w │      1.28% │
│ 20-reads (incr and recompute)      │   1_848.19ns │   2_772.00w │     0.47w │     0.47w │      1.19% │
│ 20-reads (restore from cache)      │   1_769.90ns │   2_680.00w │     0.43w │     0.43w │      1.14% │
│ clique (create and compute)        │  58_843.28ns │  69_748.00w │   348.84w │   348.84w │     38.02% │
│ clique (incr and recompute)        │  62_126.76ns │  76_504.00w │   196.04w │   196.04w │     40.14% │
│ clique (restore from cache)        │  19_829.28ns │  28_844.00w │    45.60w │    45.60w │     12.81% │
│ bipartite (create and compute)     │ 154_778.31ns │ 156_143.00w │ 2_567.61w │ 2_567.61w │    100.00% │
│ bipartite (incr and recompute)     │ 127_557.31ns │ 140_430.00w │   386.40w │   386.40w │     82.41% │
│ bipartite (restore from cache)     │  37_230.86ns │  57_997.00w │    32.26w │    32.26w │     24.05% │
│ memo diamonds (create and compute) │   9_395.45ns │  11_883.00w │    33.33w │    33.33w │      6.07% │
│ memo diamonds (incr and recompute) │  12_549.82ns │  16_375.00w │    26.41w │    26.41w │      8.11% │
│ memo diamonds (restore from cache) │   4_869.12ns │   6_507.00w │     9.36w │     9.36w │      3.15% │
└────────────────────────────────────┴──────────────┴─────────────┴───────────┴───────────┴────────────┘
```

**Review by aalekseyev — CHANGES_REQUESTED** [review]
The PR look good overall, but as we discussed there's a few changes we should make:

- It looks that there's a bug where `start_computing` will run multiple times for the same node if multiple calls to `consider_and_compute_without_adding_dep` are blocked in `of_restoring` at the same time. We should fix that. One idea is to "re-enter" `consider_and_compute_without_adding_dep` and look at the `state` again.

    I'm slightly worried about executing this logic per-edge rather than per-node, as well, but it's probably fine because the work is small. 

- It looks like we're no longer benefiting from the `Not_forced` state of `Once.t` because we only create the `Once.t` values as we're about to force them. Maybe we could get rid of that.

- I liked your idea of merging `No_value` and `Old_value` variants together.

**Comment by snowleopard** [comment]
@aalekseyev Thanks for the review and suggestions! I've addressed everything and here are some new microbenchmarks. There is some memory saving but timewise the results seem to be roughly the same (there seem to be a slight improvement). Manually benchmarking a large build shows similar results too.

```
┌────────────────────────────────────┬──────────────┬─────────────┬───────────┬───────────┬────────────┐
│ Without Once                       │     Time/Run │     mWd/Run │  mjWd/Run │  Prom/Run │ Percentage │
├────────────────────────────────────┼──────────────┼─────────────┼───────────┼───────────┼────────────┤
│ 1-bind (create and compute)        │     432.14ns │     608.00w │     0.10w │     0.10w │      0.28% │
│ 1-bind (incr and recompute)        │     450.26ns │     581.00w │           │           │      0.29% │
│ 1-bind (restore from cache)        │     364.33ns │     487.00w │           │           │      0.23% │
│ 20-reads (create and compute)      │   2_042.90ns │   3_102.00w │     0.73w │     0.73w │      1.31% │
│ 20-reads (incr and recompute)      │   1_894.97ns │   2_842.00w │     0.49w │     0.49w │      1.22% │
│ 20-reads (restore from cache)      │   1_813.52ns │   2_748.00w │     0.44w │     0.44w │      1.17% │
│ clique (create and compute)        │  60_027.11ns │  71_273.00w │   349.06w │   349.06w │     38.62% │
│ clique (incr and recompute)        │  63_498.72ns │  77_192.00w │   194.96w │   194.96w │     40.86% │
│ clique (restore from cache)        │  20_371.68ns │  30_305.00w │    45.57w │    45.57w │     13.11% │
│ bipartite (create and compute)     │ 155_412.42ns │ 158_218.00w │ 2_601.14w │ 2_601.14w │    100.00% │
│ bipartite (incr and recompute)     │ 127_554.60ns │ 141_525.00w │   388.33w │   388.33w │     82.07% │
│ bipartite (restore from cache)     │  37_392.63ns │  59_890.00w │    32.59w │    32.59w │     24.06% │
│ memo diamonds (create and compute) │   9_107.59ns │  11_458.00w │    31.24w │    31.24w │      5.86% │
│ memo diamonds (incr and recompute) │  12_043.28ns │  15_410.00w │    25.86w │    25.86w │      7.75% │
│ memo diamonds (restore from cache) │   4_649.63ns │   6_040.00w │     8.45w │     8.45w │      2.99% │
└────────────────────────────────────┴──────────────┴─────────────┴───────────┴───────────┴────────────┘
```

**Comment by snowleopard** [comment]
I also added the number of paths in the DAG to metrics. This matches the number of blocking edges we discovered.

This emphasises a problem we have in our testsuite: we never have more than one path in the DAG. It would be nice to make the tests stronger, by having at least one test with 2 paths.

File: src/memo/memo.ml [hunk 1]
```diff
@@ -747,7 +741,7 @@ end
    Below are some notes on how/why this algorithm works.
 
    By "computation" we mean execution of a "shared fiber", which we represent by
-   a [Once.t]. Such computations can be in one of three states: not started,
+   a [Computation.t]. Computations can be in one of three states: not started,
```
Comment (aalekseyev) [inline src/memo/memo.ml:744]: This comment should probably be updated: computations can no longer be "not started", I assume?

Comment (aalekseyev) [reply src/memo/memo.ml:744]: Oh, never mind, we still have `Computation.create` and `Computation.force` as separate functions. 
I guess that's a bit tricky to eliminate.

**Review by aalekseyev — APPROVED** [review]
Looks good!

File: src/memo/memo.ml (hunk 1 above)
Comment (snowleopard) [reply src/memo/memo.ml:744]: Yeah, I tried eliminating it but the code gets a bit convoluted as a result, so I left `create` separate.

**Pull request merged**